AsyncDisplayKit/Private/ASCollectionDataInternal.mm

```diff
@@ -0,0 +1,320 @@
+{
+  NSMutableDictionary<NSNumber *, ASCollectionItemImpl *> *dict = _supplementaryElements[kind];
+  if (dict == nil) {
+    dict = [NSMutableDictionary dictionary];
+    _supplementaryElements[kind] = dict;
+  }
+  ASDisplayNodeAssertNil(dict[@(index)], @"Supplementary element of kind %@ already exists at index %td. Identifier: %@", kind, index, dict[@(index)].identifier);
+  dict[@(index)] = item;
+}
+
+- (NSString *)description
+{
+  return ASObjectDescriptionMake(self, [self propertiesForDescription]);
+}
+
+- (NSMutableArray <NSDictionary *> *)propertiesForDescription
+{
+  NSMutableArray *array = [NSMutableArray array];
+  [array addObject:@{ @"identifier" : _identifier }];
+  [array addObject:@{ @"items" : _mutableItems }];
+  if (_supplementaryElements.count > 0) {
+    [array addObject:@{ @"supplementaries" : _supplementaryElements }];
+  }
+  return array;
+}
+
+@end
+
+@implementation ASCollectionData {
+  ASCollectionSectionImpl *_currentSection;
+
+  NSMutableDictionary<ASItemIdentifier, ASCollectionItemImpl *> *_itemsDict;
+  NSMutableDictionary<ASSectionIdentifier, ASCollectionSectionImpl *> *_sectionsDict;
+
+  // We could have used NSMutableSet for these, but copying the dictionary is probably faster than
+  // creating an array of all keys, then creating a set of those, and then creating an array from the set
+  // during the copy.
+  NSMutableDictionary<ASItemIdentifier, ASCollectionItemImpl *> *_usedItems;
+  NSMutableDictionary<ASSectionIdentifier, ASCollectionSectionImpl *> *_usedSections;
+
+  NSMutableSet *_supplementaryElementKinds;
+  BOOL _completed;
+}
+
+- (instancetype)initWithReusableContentFromCompletedData:(ASCollectionData *)data
+{
+  self = [super init];
+  if (self != nil) {
+    _usedItems = [NSMutableDictionary dictionary];
+    _usedSections = [NSMutableDictionary dictionary];
+    _mutableSections = [NSMutableArray array];
+    _supplementaryElementKinds = [NSMutableSet set];
+    
+    if (data != nil) {
+      ASDisplayNodeAssert(data->_completed, @"You must pass a completed collection data.");
+      _itemsDict = data->_usedItems;
+      _sectionsDict = [[NSMutableDictionary alloc] initWithDictionary:data->_usedSections copyItems:YES];
+    } else {
+      _itemsDict = [NSMutableDictionary dictionary];
+      _sectionsDict = [NSMutableDictionary dictionary];
+    }
+  }
+  return self;
+}
+
+- (instancetype)init
+{
+  return [self initWithReusableContentFromCompletedData:nil];
+}
+
+#pragma mark - Convenience Builders (Public)
+
+- (void)addSectionWithIdentifier:(ASSectionIdentifier)identifier block:(void (^)(ASCollectionData * _Nonnull))block
+{
+  if (_currentSection != nil) {
+    ASDisplayNodeFailAssert(@"Call to %@ must not be inside an addSection: block.", NSStringFromSelector(_cmd));
+    return;
+  }
+
+  _currentSection = (ASCollectionSectionImpl *)[self sectionWithIdentifier:identifier];
```

Comment:
Same question about SectionImpl, if it's possible to unify with the public Section object and only expose limited API.